llama-index-integrations/instrumentation/llama-index-instrumentation-agentops/llama_index/instrumentation/agentops/base.py

```diff
@@ -0,0 +1,236 @@
+
+
+class AgentOpsSpanHandler(SimpleSpanHandler):
+    _shared_handler_state: AgentOpsHandlerState = PrivateAttr()
+    _ao_client: AOClient = PrivateAttr()
+    _observed_exceptions: Set[Exception] = PrivateAttr()
+
+    def __init__(
+        self, shared_handler_state: AgentOpsHandlerState, ao_client: AOClient
+    ) -> None:
+        self._shared_handler_state = shared_handler_state
+        self._ao_client = ao_client
+        self._observed_exceptions = set()
+        super().__init__()
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "AgentOpsSpanHandler"
+
+    def new_span(
+        self,
+        id_: str,
+        bound_args: BoundArguments,
+        instance: Optional[Any] = None,
+        parent_span_id: Optional[str] = None,
+        **kwargs: Any
+    ) -> SimpleSpan:
+        self._shared_handler_state.is_agent_chat_span[id_] = False
+        self._shared_handler_state.span_parent[id_] = parent_span_id
+        return super().new_span(id_, bound_args, instance, parent_span_id, **kwargs)
+
+    def prepare_to_exit_span(
+        self,
+        id_: str,
+        bound_args: BoundArguments,
+        instance: Optional[Any] = None,
+        result: Optional[Any] = None,
+        **kwargs: Any
+    ) -> SimpleSpan:
+        self._shared_handler_state.is_agent_chat_span.pop(id_, None)
+        self._shared_handler_state.agent_chat_start_event.pop(id_, None)
+        self._shared_handler_state.span_parent.pop(id_, None)
+        self._shared_handler_state.span_exception.pop(id_, None)
+        return super().prepare_to_exit_span(id_, bound_args, instance, result, **kwargs)
+
+    def prepare_to_drop_span(
+        self,
+        id_: str,
+        bound_args: BoundArguments,
+        instance: Optional[Any] = None,
+        err: Optional[BaseException] = None,
+        **kwargs: Any
+    ) -> SimpleSpan:
+        if err:
+            # Associate this exception to the parent span, so that it will know to ignore it
+            parent_span_id = self._shared_handler_state.span_parent.get(id_, None)
+            if parent_span_id:
+                if parent_span_id in self._shared_handler_state.span_exception:
+                    self._shared_handler_state.span_exception[parent_span_id].add(err)
+                else:
+                    self._shared_handler_state.span_exception[parent_span_id] = {err}
+
+            # If this exception hasn't yet been associated to this span, record it
+            if (
+                id_ not in self._shared_handler_state.span_exception
+                or err not in self._shared_handler_state.span_exception[id_]
+            ):
+                self._ao_client.record(ErrorEvent(details=str(err)))
+
+        self._shared_handler_state.is_agent_chat_span.pop(id_, None)
+        self._shared_handler_state.agent_chat_start_event.pop(id_, None)
+        self._shared_handler_state.span_parent.pop(id_, None)
+        self._shared_handler_state.span_exception.pop(id_, None)
```

Comment:
Should we put all these pops in a helper function? `def _remove_span_id_from_state()` ?